### ISSUE TITLE
[FEATURE] Persistence inside of Jobs

### ISSUE THREAD

**Issue by cupcakearmy** [python-telegram-bot/python-telegram-bot#1671]
The job runner are amazing. But they lack some documentation in terms of persistence. For me it turned out to do something along the lines of this:

```python
persistence = PicklePersistence(DB_FILE)
updater: Updater = Updater('token', use_context=True, persistence=persistence)

def update_updater_data():
    updater.dispatcher.user_data = persistence.user_data
    updater.dispatcher.update_persistence()
    persistence.flush()

def my_job():
    persistence.user_data......do stuff and save it
    update_updater_data()
```

which works, but took me a lot of trial and error.
Is there a better approach to this? (maybe I'm missing something)
It might be a good idea to add a little section either in the persistence or jobqueue wiki site :)

**Comment by Bibo-Joshi** [comment]
Hi @cupcakearmy. Haven't had too close of a look at your code yet, but wanted to point you [here](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Making-your-bot-persistent#what-can-become-persistent). Also see the page linked therein.

**Comment by cupcakearmy** [comment]
@Bibo-Joshi thanks for the suggestion, but that is not the use case I'm searching for. I basically need the user/chat data inside the JobQueue

**Comment by Eldinnie** [comment]
well, since the dispatcher is part of the context you get passed in a job, you can have it :D

**Comment by Bibo-Joshi** [comment]
I think your main problem is, that `user/chat_data` is only persisted on handler execution and not on *job* execution, isn't it?
IMO that would indeed be a good enhancement.

**Comment by cupcakearmy** [comment]
The most confusing (and infuriating) part was that in the docs there is not much written about how one should go about persistence in jobs. This led to me trying to modify and flush the persistence variable at first and wondering why it would not work.

```python
persistence = PicklePersistence(DB_FILE)
updater: Updater = Updater(token, use_context=True, persistence=persistence)
```

That did not work because the updater keeps a local copy of the data. Only by navigating through the source code I later discovered that I basically need to update the data inside the dispatcher and not the persistence since the dispatcher will overwrite anything you have written to it from its local state.

It took me longer that it I think it should, as the solution is actually very simple once I started going through source code.

So basically what I as a neewbe to the package would have loved to have a clear documentation would have been:

1. how do I access chat_data & user_data from within jobs.
2. How to persist them in jobs if I made any changes.

**Comment by Bibo-Joshi** [comment]
Mh, I guess we're basically on the same page here, the issue being data not being persisted in jobs.

From my point of view missing documentation is not really a problem here. Accessing is as simple as `context.user/chat_data` and documented [here](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.callbackcontext.html) and [here](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Storing-user--and-chat-related-data).
The info about how persistence is handled internally is nothing for the docstring. You needed that info because you encountered a bug you wanted to solve, i.e. there is no doc about persisting in jobs because we didn't think of that case yet. But with that bug fixed the normal user won't need to fiddle with the persistence manually.

Would you maybe like to pr yourself to make sure data is persisted after job execution?

**Comment by cupcakearmy** [comment]
Sry for the inactivity! I was sure I had replied :/

I could see if I manage to fix it in the source code, but no promises :)

**Comment by Bibo-Joshi** [comment]
No problem.
Sweet, please go ahead then :) If you have any questions, don't hesitate to ask!
